src/r1cs/mod.rs

```diff
@@ -159,6 +160,76 @@ impl<E: Engine> R1CSShape<E> {
     })
   }
 
+  /// Generate a random [R1CSShape] with the specified number of constraints, variables, and public inputs/outputs.
+  pub fn random<R: RngCore + CryptoRng>(
+    num_cons: usize,
+    num_vars: usize,
+    num_io: usize,
+    num_entries: usize,
+    mut rng: &mut R,
+  ) -> Self {
+    assert!(num_cons.is_power_of_two());
+    assert!(num_vars.is_power_of_two());
+    assert!(num_entries.is_power_of_two());
+    assert!(num_io < num_vars);
+
+    let num_rows = num_cons;
+    let num_cols = num_vars + 1 + num_io;
+
+    let (NA, NB, NC) = {
+      let N_div_3 = num_entries / 3;
+      let NC = num_entries - (2 * N_div_3);
+      (N_div_3, N_div_3, NC)
+    };
+
+    let A = SparseMatrix::random(num_rows, num_cols, NA, &mut rng);
+    let B = SparseMatrix::random(num_rows, num_cols, NB, &mut rng);
+    let C = SparseMatrix::random(num_rows, num_cols, NC, &mut rng);
+
+    Self {
+      num_cons,
+      num_vars,
+      num_io,
+      A,
+      B,
+      C,
+      digest: Default::default(),
+    }
+  }
+
+  /// Generate a satisfying [RelaxedR1CSWitness] and [RelaxedR1CSInstance] for this [R1CSShape].
```

Comment:
We might want to signal in a comment if the witness satisfies the instance, as this is relevant to the range of tests (positive, negative) one can write with this.

Reply:
Do you mean this could be interpreted as "the witness is satisfying but the instance may not"? Or should the method name be clearer?